orchagent/pfcqueue.cpp

```diff
@@ -0,0 +1,394 @@
+            finalStats[1] - m_stats[1]);
+}
+
+std::vector<uint64_t> PfcQueue::getQueueStats(void)
+{
+    SWSS_LOG_ENTER();
+
+    std::vector<sai_queue_stat_t> queueStatIds =
+    {
+        SAI_QUEUE_STAT_PACKETS,
+        SAI_QUEUE_STAT_DROPPED_PACKETS,
+    };
+
+    std::vector<uint64_t> queueStats(queueStatIds.size(), 0);
+
+    sai_status_t status = sai_queue_api->get_queue_stats(
+            m_queue,
+            static_cast<uint32_t>(queueStatIds.size()),
+            queueStatIds.data(),
+            queueStats.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get queue 0x%lx stats: %d", m_queue, status);
+    }
+
+    return std::move(queueStats);
+}
+
+PfcLossyQueue::PfcLossyQueue(sai_object_id_t port, sai_object_id_t queue, uint32_t queueId):
+    PfcQueue(port, queue, queueId)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+
+    sai_status_t status = sai_port_api->get_port_attribute(port, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC mask on port 0x%lx: %d", port, status);
+    }
+
+    uint8_t pfcMask = attr.value.u8;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+    attr.value.u8 = pfcMask & ~(1 << queueId);
+
+    status = sai_port_api->set_port_attribute(port, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC mask on port 0x%lx: %d", port, status);
+    }
+}
+
+PfcLossyQueue::~PfcLossyQueue(void)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+
+    sai_status_t status = sai_port_api->get_port_attribute(port(), 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC mask on port 0x%lx: %d", port(), status);
+        throw std::runtime_error("PfcLossyQueue initialization failure");
+    }
+
+    uint8_t pfcMask = attr.value.u8;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+    attr.value.u8 = pfcMask | (1 << queueId());
+
+    status = sai_port_api->set_port_attribute(port(), &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set PFC mask on port 0x%lx: %d", port(), status);
+        throw std::runtime_error("PfcLossyQueue initialization failure");
+    }
+}
+
+PfcZeroBufferQueue::PfcZeroBufferQueue(sai_object_id_t port,
+        sai_object_id_t queue, uint32_t queueId):
+    PfcLossyQueue(port, queue, queueId)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_QUEUE_ATTR_BUFFER_PROFILE_ID;
+
+    // Get queue's buffer profile ID
+    sai_status_t status = sai_queue_api->get_queue_attribute(queue, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get buffer profile ID on queue 0x%lx: %d", queue, status);
+        throw std::runtime_error("PfcZeroBufferQueue initialization failure");
```

Comment:
Not sure throwing exception is a good solution.

Reply:
done